src/com/amazon/ion/impl/IonReaderLookaheadBuffer.java

```diff
@@ -0,0 +1,872 @@
+            // to skip than are currently available (e.g. if a header or trailer is incomplete).
+            return 0;
+        }
+    }
+
+    /**
+     * Retrieve and buffer up to {@link #pageSize} bytes from the input.
+     * @param numberOfBytesRequested the minimum amount of space that must be available before the buffer reaches
+     *                               its configured maximum size.
+     * @return the number of bytes buffered by this operation.
+     * @throws Exception if thrown by the underlying InputStream.
+     */
+    private int fillPage(int numberOfBytesRequested) throws Exception {
+        int amountToFill = pipe.capacity() - pipe.size();
+        if (amountToFill <= 0) {
+            // Try to fill the remainder of the existing buffer to avoid growing unnecessarily. If there is no
+            // space, that indicates that a single value exceeds the size of a page. In that case, fill another page,
+            // growing the buffer only up to the configured maximum size.
+            int spaceAvailable = getMaximumBufferSize() - pipe.capacity();
+            if (numberOfBytesRequested > spaceAvailable) {
+                if (nopPadStartIndex > -1 && valuePreHeaderIndex - nopPadStartIndex >= numberOfBytesRequested) {
+                    // Reclaim the NOP pad space if doing so would allow the value to fit.
+                    reclaimNopPadding();
+                } else {
+                    startSkippingValue();
+                }
+                amountToFill = numberOfBytesRequested;
+            } else {
+                amountToFill = Math.min(pageSize, spaceAvailable);
+            }
+        }
+        int received;
+        if (isSkippingCurrentValue()) {
+            if (state == State.SKIPPING_VALUE) {
+                // This is a seek operation, meaning that the bytes don't need to be interpreted.
+                received = skipBytesFromInput(amountToFill);
+            } else {
+                // The bytes need to be interpreted, so they cannot be skipped. The caller must retrieve them from
+                // the input.
+                received = amountToFill;
+            }
+        } else {
+            received = pipe.receive(getInput(), amountToFill);
+        }
+        return received;
+    }
+
+    /**
+     * Notifies the event handler that the current value is oversized, if the handler has not already been notified.
+     * @throws Exception if thrown by the handler.
+     */
+    private void notifyHandlerOfOversizedValue() throws Exception {
+        if (handlerNeedsToBeNotifiedOfOversizedValue && eventHandler != null) {
+            if (isSystemValue) {
+                // Processing cannot continue after system values (symbol tables) are truncated because subsequent
+                // values may be unreadable. Notify the user.
+                ((IonBufferEventHandler) eventHandler).onOversizedSymbolTable();
```

Comment:
It seems like the `IonReaderLookaheadBuffer` could reasonably use an `IonBufferEventHandler` directly. What's the motivation for this downcast?

Reply:
Note: after splitting the EventHandler into individual functional interfaces, this is now moot.